Caboodle/Permissions/Permissions.android.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Android;
+using Android.Content.PM;
+using Android.Support.V4.App;
+using Android.Support.V4.Content;
+
+namespace Microsoft.Caboodle
+{
+    internal static partial class Permissions
+    {
+        static TaskCompletionSource<PermissionStatus> tcs;
+
+        static void PlatformEnsureDeclared(PermissionType permission)
+        {
+            var androidPermissions = permission.ToAndroidPermissions();
+
+            // No actual android permissions required here, just return
+            if (androidPermissions == null || !androidPermissions.Any())
+                return;
+
+            var context = Platform.CurrentContext;
+
+            foreach (var ap in androidPermissions)
+            {
+                var packageInfo = context.PackageManager.GetPackageInfo(context.PackageName, PackageInfoFlags.Permissions);
+                var requestedPermissions = packageInfo?.RequestedPermissions;
+
+                // If the manifest is missing any of the permissions we need, throw
+                if (!requestedPermissions?.Any(r => r.Equals(ap, StringComparison.OrdinalIgnoreCase)) ?? false)
+                    throw new PermissionException($"You need to declare the permission: `{ap}` in your AndroidManifest.xml");
+            }
+        }
+
+        static Task<PermissionStatus> PlatformCheckStatusAsync(PermissionType permission)
+        {
+            PlatformEnsureDeclared(permission);
+
+            var androidPermissions = permission.ToAndroidPermissions();
+
+            if (androidPermissions == null || !androidPermissions.Any())
+                return Task.FromResult(PermissionStatus.Unknown);
+
+            var hasApiM = Platform.HasApiLevel(Android.OS.BuildVersionCodes.M);
+            var context = Platform.CurrentContext;
+
+            foreach (var ap in androidPermissions)
+            {
+                if (hasApiM)
```

Comment:
@jamesmontemagno is this truly necessary? Shouldn't `PermissionChecker` handle this for us regardless of API level?

Reply:
You actually need to check if it is TARGETING... not running on M+

```
	var targetsMOrHigher = context.ApplicationInfo.TargetSdkVersion >= Android.OS.BuildVersionCodes.M;

			foreach (var name in names)
			{
				if (targetsMOrHigher)
				{
					if (ContextCompat.CheckSelfPermission(context, name) != Android.Content.PM.Permission.Granted)
						return Task.FromResult(PermissionStatus.Denied);
				}
				else
				{
					if(PermissionChecker.CheckSelfPermission(context, name)  != PermissionChecker.PermissionGranted)
						return Task.FromResult(PermissionStatus.Denied);
				}
            }
```

Reply:
Oops, missed that... However my question remains, why do we need to change code paths when support libraries should do this dirty work for us already?

Reply:
It doesn't :( trust me, it sucks.